### PR TITLE
core/mr_cache: Add access checks to mr cache

### PR DESCRIPTION
**Pull request by shefty**

### CODE REVIEW

**Comment by shefty** [comment]
The Intel CI error looks like some sort of node configuration problem.  Failure is OMPI over sockets not starting.

**Comment by shefty** [comment]
Added patch to verify access permissions as part of cache search

File: prov/verbs/src/verbs_mr.c [hunk 1]
```diff
@@ -215,10 +215,8 @@ int fi_ibv_mr_cache_add_region(struct ofi_mr_cache *cache,
 	md->mr_fid.fid.ops = &fi_ibv_mr_cache_fi_ops;
 	md->entry = entry;
 
-	return fi_ibv_mr_reg_common(md, IBV_ACCESS_LOCAL_WRITE |
-			IBV_ACCESS_REMOTE_WRITE | IBV_ACCESS_REMOTE_ATOMIC |
-			IBV_ACCESS_REMOTE_READ, entry->info.iov.iov_base,
-			entry->info.iov.iov_len, NULL);
+	return fi_ibv_mr_reg_common(md, ofi_mr_access_map[entry->info.mr_access],
```
Comment (shefty) [inline prov/verbs/src/verbs_mr.c]: This is wrong.  ofi_mr_access_map is returning FI_xxx permissions (FI_READ, FI_SEND, etc.), not libibverbs access.

**Comment by shefty** [comment]
CI failure is:
```
IMB-MPI1: prov/util/src/util_mr_cache.c:430: ofi_mr_rbt_erase: Assertion `node' failed.
```
I'm not sure of the exact reason for this error, but I believe there's an issue in the util_mr_uncache_entry_storage() path.  That call should remove 'entry' from the underlying rbtree.  The call stack is
```
util_mr_uncache_entry_storage()
cache->storage.erase()
ofi_mr_rbt_erase()
ofi_rbmap_find()
map->compare()
util_mr_find_within() or util_mr_find_overlap()
```
In general, there's no guarantee that the compare() callback will find the entry corresponding to the exact entry that is being uncached.  This could happen when normal cache insertions are made using find_within(), but memory notifications must match using find_overlap().

I believe we need some way to have the erase path remove exactly the entry being uncached.  If we didn't have a generic storage class, it would be easier.

**Comment by shefty** [comment]
I backed off changes to this PR while I work on other fixes.  Only the basic prep work is here.

@rajachan - Does the AWS failure make sense to you?  Is it related to this change?

**Comment by rajachan** [comment]
The fi_close() warnings are an artifact of an issue we are tracking internally with the dgram endpoint's MR cleanup path. I do not expect the test to timeout though. I'll try your patch manually in a little while.

**Comment by rajachan** [comment]
I have not been able to reproduce the timeout with your patch, might have been a side-effect of the known issue I mentioned. I'll look at it more closely, but safe to say the failure is unrelated to your patch.

**Comment by shefty** [comment]
@rajachan - thanks for checking

**Pull request merged**